Drag n drop sorting for custom fields

Kovah opened the pull request:
This pull request implements the ability to sort custom fields via drag n drop while editing them. The order is preserved when saving the entry.
The drag n drop feature uses the Angular CDK DragDropModule to provide frictionless support.

Preview:
![bitwarden_sorting_preview](https://user-images.githubusercontent.com/1816101/55037694-e0aa2a00-501e-11e9-9d48-c8c77777292e.gif)

I had to move the styles for the `.box-content-row` class out of the `.box` declaration to prevent hundreds of lines code duplication. I hope this won't be a problem but I couldn't find any conflicting styles.

:warning: This pull request has a dependent pull request in the bitwarden/jslib repository which needs to be merged first before this pull request can be merged. Once it's merged I will update the submodule reference.

kspearrin commented:
I am still confused about why there are so many style changes.

kspearrin commented on src/app/vault/add-edit.component.html:
```diff
@@ -261,36 +261,42 @@
+                                <span class="box-drag-handle" title="{{'drag' | i18n}}" cdkDragHandle>
+                                    <i class="fa fa-bars"></i>
+                                </span>
+                                <a href="#" class="box-content-delete" appStopClick (click)="removeField(f)"
```
i dont see this `box-content-delete` class being used anywhere. why was it added?

kspearrin commented on src/app/vault/add-edit.component.html:
```diff
@@ -261,36 +261,42 @@
-                                    title="{{'toggleVisibility' | i18n}}" (click)="toggleFieldValue(f)">
-                                    <i class="fa fa-lg"
-                                        [ngClass]="{'fa-eye': !f.showValue, 'fa-eye-slash': f.showValue}"></i>
+                    <div cdkDropList (cdkDropListDropped)="drop($event)">
```
Can you explain what changed in this block? Having a hard time seeing the diff here. Looks like a new wrapping div and the `box-drag-handle` span. Is that all?

kspearrin commented on src/scss/box.scss:
```diff
@@ -1,6 +1,7 @@
 @import "variables.scss";
```
confused about why so much change

kspearrin commented on src/app/vault/add-edit.component.html:
```diff
@@ -261,36 +261,42 @@
+                            <div class="box-content-row box-content-row-multi box-draggable-row" cdkDrag
+                                *ngFor="let f of cipher.fields; let i = index; trackBy:trackByFunction"
+                                [ngClass]="{'box-content-row-checkbox': f.type === fieldType.Boolean}">
+                                <span class="box-drag-handle" title="{{'drag' | i18n}}" cdkDragHandle>
```
does a locale entry for `drag` exist?

Kovah commented:
Here's an overview on the style and HTML changes:

* Wrapped the `ng-container` with a div that handles the cdkDropList, to keep the "Add Field" box out of the sorting list.
* Added a span that works as the drag handle. (Will remove the box-content-delete, used it for styling but works without it)
* Moved the whole styles for `.box-content-row` out of the `.box` block (maybe the reason for all those changes in the styles file), but there are no style changes for the box-content-row itself.
* Added a new `.box-drag-handle` block for styling the drag handle.
* Added a new `.cdk-drag-preview` block for styling the moving custom field box.

A locale entry for `drag` does not exist yet, will add it in a minute.

kspearrin commented:
Thanks for the details. But, I still don't understand why `.box-content-row` needs to be outside of `.box` now. Is it because the element shown when dragging it around is outside of that scope?

Kovah commented:
Correct. The dragged field is copied/moved to the end of the </body> tag and thus loses all styling of the surrounding .box element. Moving the styles out of the scope solves this and prevents duplicate styles.

kspearrin commented:
Ok. I'm not a big fan of the placement of the drag handles in the current screenshot since the minus icons are now misaligned with the "add new field" plus. I might play around with moving it so that the icons can still line up.

Are you able to add this to the browser and web projects as well?

Kovah commented:
Sure, no problem. Handles might be placed on the right side or wherever it makes sense.

> Are you able to add this to the browser and web projects as well?

Sure, will take a look at it as soon as possible.

Kovah commented:
Just played with the placement of the handle, looks best the far right side of the field box, in my opinion. Would use this for all implementations if you are okay with it.

![Screenshot 2019-03-27 at 15 22 24](https://user-images.githubusercontent.com/1816101/55083660-2bba5080-50a4-11e9-8b63-3e86646a2956.jpg)

kspearrin commented:
yea i think i like that better

kspearrin commented on package-lock.json:
```diff
@@ -1397,12 +1406,20 @@
-      "version": "2.6.1",
-      "resolved": "https://registry.npmjs.org/async/-/async-2.6.1.tgz",
-      "integrity": "sha512-fNEiL2+AZt6AlAw/29Cr0UDe4sRAHCpEHh54WMz+Bb7QfNcFw4h3loofyJpLeQs4Yx7yuqu/2dLgM5hKOs6HlQ==",
+      "version": "2.6.2",
```
Did you run some other npm updates here? Not sure why this package was bumped since it doesn't seem to be a dependency of cdk.

Kovah replied:
NPM seems to update packages on its own despite running npm install only. Should I revert this?

kspearrin replied:
Yes, please only make changes that are necessary for this PR.
